### PR TITLE
[new module] azure_rm_keyvault

### PR DESCRIPTION
**Pull request by zikalino**
##### SUMMARY
Adding support for azure_rm_keyvault

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME

azure_rm_keyvault

##### ANSIBLE VERSION

 2.4

### CODE REVIEW

**Comment by zikalino** [comment]
@jborean please check this pr, pls note fix for sanity test is in a separate pr

**Comment by mattclay** [comment]
CI failure in integration tests due to traceback:

```
Traceback (most recent call last):
  File "/tmp/ansible_i9pK4R/ansible_module_azure_rm_keyvault.py", line 435, in <module>
    main()
  File "/tmp/ansible_i9pK4R/ansible_module_azure_rm_keyvault.py", line 432, in main
    AzureRMVaults()
  File "/tmp/ansible_i9pK4R/ansible_module_azure_rm_keyvault.py", line 265, in __init__
    supports_tags=False)
  File "/tmp/ansible_i9pK4R/ansible_modlib.zip/ansible/module_utils/azure_rm_common.py", line 321, in __init__
  File "/tmp/ansible_i9pK4R/ansible_module_azure_rm_keyvault.py", line 307, in exec_module
    self.adjust_parameters()
AttributeError: 'AzureRMVaults' object has no attribute 'adjust_parameters'
```

File: lib/ansible/modules/cloud/azure/azure_rm_keyvault.py [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+            - The name of the Resource Group to which the server belongs.
+        required: True
+    vault_name:
+        description:
+            - Name of the vault
+        required: True
+    location:
+        description:
+            - Resource location. If not set, location from the resource group will be used as default.
+    tenant_id:
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_keyvault.py]: change this to `vault_tenant`. `tenant_id` is too close to the base option `tenant` used as a credential arg.

File: lib/ansible/modules/cloud/azure/azure_rm_keyvault.py [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+            name:
+                description:
+                    - SKU name to specify whether the key vault is a standard vault or a premium vault.
+                required: True
+                choices:
+                    - 'standard'
+                    - 'premium'
+    access_policies:
+        description:
+            - "An array of 0 to 16 identities that have access to the key vault. All identities in the array must use the same tenant ID as the key vault's t
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_keyvault.py]: don't put a newline right in the middle of a work, just do

```
description:
- An array of 0 to 16 identities that have access to the key value. All identities in the array must use the same
  tenant ID as the key vault's tenant ID.
```

File: lib/ansible/modules/cloud/azure/azure_rm_keyvault.py [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+    enable_soft_delete:
+        description:
+            - Property to specify whether the soft delete functionality is enabled for this key vault.
+        type: bool
+    create_mode:
+        description:
+            - "The vault's create mode to indicate whether the vault need to be recovered or not."
+        choices:
+            - 'recover'
+            - 'default'
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_keyvault.py]: Should be set as the default explicitly

File: lib/ansible/modules/cloud/azure/azure_rm_keyvault.py [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+            if self.state == 'absent':
+                self.to_do = Actions.Delete
+            elif self.state == 'present':
+                self.log("Need to check if Key Vault instance has to be deleted or may be updated")
+                self.to_do = Actions.Update
+
+        if (self.to_do == Actions.Create) or (self.to_do == Actions.Update):
+            self.log("Need to Create / Update the Key Vault instance")
+
+            if self.check_mode:
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_keyvault.py:397]: Running this in check mode will always result in a change even when an update is not needed. You need to do check the existing object to see if it needs updating before you exit here.

**Review by jborean93 — CHANGES_REQUESTED** [review]

**Comment by zikalino** [comment]
@jborean93 pls recheck and possibly merge

**Comment by zikalino** [comment]
@nitzmahone could you pls check this pr and possibly merge?

**Comment by zikalino** [comment]
@jborean93 just made changes as requested by @nitzmahone today.
would you be able to merge this module?

**Comment by kyliel** [comment]
@nitzmahone , any other feedback to move forward? Thanks.

**Comment by nitzmahone** [comment]
@zikalino The 2.5.0a1 release stuff took a lot longer than I planned today (we're changing up a bunch of the release processes), so I'll get back to this on Monday and get it merged.

**Comment by nitzmahone** [comment]
@zikalino Update was totally broken, so had to spend some time fixing that up and added a really basic test for it. Also, "current tenant" doesn't appear to be easily available from CLI or UserPassCredential, so I had to back out the automatic vault tenant_id change. :(

I ended up not `set`-ifying the access_policies list, as the way you're comparing values from the response dict and the Ansible module dict would likely cause issues (eg, missing key vs key present and value is `None`). We wouldn't have this issue with the SDK objects, as they'd normalize this difference automatically (and provide us hashable objects, too). So maybe chalk that up as one more plus for using the SDK objects.

Just as another note: the repeated `setdefault` calls makes the code really hard to read IMO, and isn't really necessary (just maybe do it once upfront for `properties` instead of on every call).

**Pull request merged**